Serial dma

janm012012 opened the pull request:
Pull-In Verspamans serial-dma branch to participate from all the fixes and optimizations he did.

hydra commented:
@janm012012 Hiya,  can I ask what your plans for merging this is?  I'm in the process of updating my CHMT48VB machine to OpenPnP and there's some features in here that would be great for me.

janm012012 commented:
Hi Dominic!
Yes, feel free to ask! And, yes, I will merge this in. However, there are a few things that I don't like to merge like ultra high serial speeds which only work after modifying the hardware or automatic speed reconfiguration while wiggling the dragpin for release or dma operation at all (its required for higher serial speeds but might add extra jitter). I've reviewed the code (locally) and reverted some of its changes but did not had the time to test is thoroughly to actually merge it. Unfortunately I currently have only very limited time to finish this up. If you like, I can probably push my changes or build a binary for you to test. If you're willing to modify your hardware as Micael has done, you may also use this version directly. IIRC he is using it in production already for some time, so its quite stable.

Jan

vespaman commented:
Hi guys, the idea with the serial setup in my code, was that it ought to be usable also on standard chmt, by setting the serial to a lower pace and disabling rts in the config file. Not sure if this has been tested though (I have not).
Are you (@hydra) intending to connect to your chmt over the existing rs485 serial interface?
There is (if you have not seen this) some more information on my github repo, on the serial interface options.

Yes I have been running this code for many hundreds of boards by now.
I have the intention to try to find a solution for the speed/acceleration change (or to remove it all together) for setups where Jan's optimization for slow serial interfaces is enabled. But I have been swamped by (unrelated) work the last couple of months.

 - Micael

hydra commented:
> Hi guys, the idea with the serial setup in my code, was that it ought to be usable also on standard chmt, by setting the serial to a lower pace and disabling rts in the config file. Not sure if this has been tested though (I have not). Are you (@hydra) intending to connect to your chmt over the existing rs485 serial interface? There is (if you have not seen this) some more information on my github repo, on the serial interface options.

@vespaman Hi!  I have 2 controller boards, one which I plan to leave with the CHMT software on so that I have a spare and/or if I sell the machine it can go with it, the other I have already flashed Jan's firmware to for testing.  One of the guys from the OpenPnP group sent me a custom USB->RS232 and USB->RS485 which has an SIL2105 IC on it and a single USB port. It exposes 2 COM ports, both of which work in smoothieware with putty and openpnp.  I'm using that for now. I can't find who sent me the board, been digging through old emails from 5 years ago but haven't found them yet.

Yes, I've seen your documentation regarding the serial IO and I raised a couple of issue regarding the incorrect refdes in the docs, I'll order up the 1MB/sec capable ICs that you mention, they're in an LCSC shopping cart...  However, rather than modifying the  existing serial ports, I've been looking at the board and notice it has a second un-populated RS422 connector, which I'm wondering if I can fully-populate and then have a serial port with flow control there.  See U13/14/15 + E12/13.  If either of you two have schematics for the board please post them somewhere for me.

My goal is to have a more reliable and flexible setup with OpenPNP than the 48VB currently gives me.

> Hi Dominic! Yes, feel free to ask! And, yes, I will merge this in. However, there are a few things that I don't like to merge like ultra high serial speeds which only work after modifying the hardware or automatic speed reconfiguration while wiggling the dragpin for release or dma operation at all (its required for higher serial speeds but might add extra jitter). I've reviewed the code (locally) and reverted some of its changes but did not had the time to test is thoroughly to actually merge it. Unfortunately I currently have only very limited time to finish this up. If you like, I can probably push my changes or build a binary for you to test. If you're willing to modify your hardware as Micael has done, you may also use this version directly. IIRC he is using it in production already for some time, so its quite stable.

@janm012012 Hi!  I'm not super concerned with ultra-high-speed serial ports, but a serial port with flow control and few communication errors appeals to me greatly.  The anti-stick-wiggle also appeals to me as some of the components in paper tapes that I use do require use of the original CHMT firmware's wiggle unstick sometimes so I prefer not to loose that feature, it's bad enough having to babysit the machine when in use already as I'm sure you guys know all too well!

Yes, if you have a more up-to-date branch for me to test I can give that a try here.

If either of you have up-to-date `machine.xml` files for me to get started with that would be awesome too.

This is the spare board I have, for Smoothieware:

https://photos.google.com/share/AF1QipNPkstrWm_m5gCZ_G7WGKSE7DTAzwro2lr-s6zTfocEpqrgPhm8PPyqekb_inxRCA/photo/AF1QipMdIp0FgPIFWwkf1odCMFnmqrSP5ss4oA7npTYP?key=NlJpY0loYjNNSFV6aVVuODVtOUlTU0RCRl82T0pn

I plan to open my machine today and swap out the cameras and drag feeders.

vespaman commented:
> One of the guys from the OpenPnP group sent me a custom USB->RS232 and USB->RS485 which has an SIL2105 IC on it and a single USB port. It exposes 2 COM ports, both of which work in smoothieware with putty and openpnp. I'm using that for now. I can't find who sent me the board, been digging through old emails from 5 years ago but haven't found them yet.

Most, if not all USB-serial bridges works with OpenPnP, as long as it is supported on the host operating system. The Exar part that I  like is a bit faster though, since it does not buffer from serial rx. But if you find this extra delay disturbing in the future, you can always change the bridge then. I'm sure you'll be fine with the 2105. (I tried 5-6 different brands/parts before realizing that the Exar part was far superior).

> 
> Yes, I've seen your documentation regarding the serial IO and I raised a couple of issue regarding the incorrect refdes in the docs,

Thanks! You are obviously correct, guess I am half blind after all :-) I'll update my docs. 

> I'll order up the 1MB/sec capable ICs that you mention, they're in an LCSC shopping cart... However, rather than modifying the existing serial ports, I've been looking at the board and notice it has a second un-populated RS422 connector, which I'm wondering if I can fully-populate and then have a serial port with flow control there. See U13/14/15 + E12/13. If either of you two have schematics for the board please post them somewhere for me.

Not sure if I follow; the RS422 (U13/U14/U15) interface, they are either UART4 rx/tx, are you thinking to use these with PE12 and PE13 for flow control? Or what do you mean with E12/13. If you want to use hardware support for cts, you would then need to live without encoder support forever. (Not that we are currently supporting encoder/closed loop, but who knows... :) OR are you thinking of using USART3 rx/tx on these pins? Then there are other options, but still not optimal, since you will have to sacrifice something else (e.g. the serial flash) ).

It was some time now, since I did the serial rewrite, but if I remember correctly, I settled for hardware controlled CTS and software controlled RTS, since the USB-serial bridges where handling the RTS signal differently, many violating the orignal rs232 standard.

Is it worth the hassle to find another serial port? Maybe you want to have the CAN driver (U18) option available for the future?


Have you found the excel sheet and x-ray of the board? That is as close to a schematic that I think we have seen so far. 

With respect to machine.xml, I suppose I might have the closest to what you want, given that I also have a 48vb, esp with regards to the extra (right hand) feeders, and if you are going to use some of my firmware features (eg the local drag pin management). I think I based my xml from Jan's, but there's a lot of different ways to solve stuff in OpenPnP, so your mileage may vary. But if you stay closer to Jan's current firmware, maybe it is better to stay with his xml.

hydra commented:
> > One of the guys from the OpenPnP group sent me a custom USB->RS232 and USB->RS485 which has an SIL2105 IC on it and a single USB port. It exposes 2 COM ports, both of which work in smoothieware with putty and openpnp. I'm using that for now. I can't find who sent me the board, been digging through old emails from 5 years ago but haven't found them yet.
> 
> Most, if not all USB-serial bridges works with OpenPnP, as long as it is supported on the host operating system. The Exar part that I like is a bit faster though, since it does not buffer from serial rx. But if you find this extra delay disturbing in the future, you can always change the bridge then. I'm sure you'll be fine with the 2105. (I tried 5-6 different brands/parts before realizing that the Exar part was far superior).

yes, I'll probably stay with the SIL2105 for now.  Noted regarding the Exar part.

> > I'll order up the 1MB/sec capable ICs that you mention, they're in an LCSC shopping cart... However, rather than modifying the existing serial ports, I've been looking at the board and notice it has a second un-populated RS422 connector, which I'm wondering if I can fully-populate and then have a serial port with flow control there. See U13/14/15 + E12/13. If either of you two have schematics for the board please post them somewhere for me.
> 
> Not sure if I follow; the RS422 (U13/U14/U15) interface, they are either UART4 rx/tx, are you thinking to use these with PE12 and PE13 for flow control? Or what do you mean with E12/13. If you want to use hardware support for cts, you would then need to live without encoder support forever. (Not that we are currently supporting encoder/closed loop, but who knows... :) OR are you thinking of using USART3 rx/tx on these pins? Then there are other options, but still not optimal, since you will have to sacrifice something else (e.g. the serial flash) ).
> Is it worth the hassle to find another serial port? Maybe you want to have the CAN driver (U18) option available for the future?

Currently I don't know, I need the schematics and MCU pinout and cross-reference that with the existing use of serial ports in Smoothieware and the MCU datasheets to see.  I was hoping to leave the existing circuits alone and just add to the board.  I don't recall which serial ports on the F4 support and have pins for RTS/CTS, again need the schematic.

> Have you found the excel sheet and x-ray of the board? That is as close to a schematic that I think we have seen so far.

I recall seeing it years ago, but can't find it now, do you have a link to it?

> With respect to machine.xml, I suppose I might have the closest to what you want, given that I also have a 48vb, esp with regards to the extra (right hand) feeders, and if you are going to use some of my firmware features (eg the local drag pin management). I think I based my xml from Jan's, but there's a lot of different ways to solve stuff in OpenPnP, so your mileage may vary. But if you stay closer to Jan's current firmware, maybe it is better to stay with his xml.

yes, I would like a machine.xml which has config for the right hand feeders, and also for the drag pin management differences of your firmware.  If you have yours that would be most helpful.

vespaman commented:
> > Have you found the excel sheet and x-ray of the board? That is as close to a schematic that I think we have seen so far.
> 
> I recall seeing it years ago, but can't find it now, do you have a link to it?

I _think_ this is the most complete, but there are other variants of it, so ... 
https://github.com/sparkfunX/Desktop-PickAndPlace-CHMT36VA/blob/master/Reverse-Engineering/chmt-pins.csv

> 
> > With respect to machine.xml, I suppose I might have the closest to what you want, given that I also have a 48vb, esp with regards to the extra (right hand) feeders, and if you are going to use some of my firmware features (eg the local drag pin management). I think I based my xml from Jan's, but there's a lot of different ways to solve stuff in OpenPnP, so your mileage may vary. But if you stay closer to Jan's current firmware, maybe it is better to stay with his xml.
> 
> yes, I would like a machine.xml which has config for the right hand feeders, and also for the drag pin management differences of your firmware. If you have yours that would be most helpful.

I'll see if I can check my machine.xml in tomorrow, and also update the component errors you found at the same time!

janm012012 commented:
Hi Guys!

> > > One of the guys from the OpenPnP group sent me a custom USB->RS232 and USB->RS485 which has an SIL2105 IC on it and a single USB port. It exposes 2 COM ports, both of which work in smoothieware with putty and openpnp. I'm using that for now. I can't find who sent me the board, been digging through old emails from 5 years ago but haven't found them yet.
> > 
> > 
> > Most, if not all USB-serial bridges works with OpenPnP, as long as it is supported on the host operating system. The Exar part that I like is a bit faster though, since it does not buffer from serial rx. But if you find this extra delay disturbing in the future, you can always change the bridge then. I'm sure you'll be fine with the 2105. (I tried 5-6 different brands/parts before realizing that the Exar part was far superior).
> 
> yes, I'll probably stay with the SIL2105 for now. Noted regarding the Exar part.
> 
I'm happy to discuss this USB issues with you. However, it might be advantageous to move over the list so that others can join in and the outcome is indexed.
For the 36VA, the serial port is isolated and requires external power. Therefore a bridge with USB-supplied 5V is recommended as it allows to remove the camera switcher board (suppose a camera upgrade is planned). In addition, the major problem IMHO between USB and serial is that USB is packet oriented while serial is character oriented. For transmission (USB -> serial) thats don't care as the bridge can immediately output all data it has. A fifo is recommended to bridge the gap before the next packet arrives. On the receive side (serial -> USB) the situation is much more complicated as for good throughput data shall be buffered and for low latency a end-of-message indication is needed. FTDI and IMHO the USB-ACM specification use timeouts and trigger character for that. An additional but related problem is given by the host-oriented nature of the USB: the host controller polls the bridge for new data and the polling rate can not be faster then 1ms. IIRC the generic Windows drivers allows to configure this interval and defaults to 10ms. I have not yet studied the datasheets of the bridges you mentioned but I would recommend one that supports the trigger-character and use a rather large timeout. Reviewing/changing the firmware such, that there is only a single character, indicating the end of the message, is on my list... (Some time back I remember having seen a packet in the sniffer send to the bridge when OpenPnP starts that was decoded as some sort of configuration message for this end-of-message/flush buffer character. Unfortunately, IIRC, is was empty and jSerial (the serial library OpenPnP uses) does not mention any API to configure it...)

Concerning the machine.xml: there are a few optimization, that are not yet in the published version:

- use Euclidean Axis Limits as motion type to provide maximum speed with minimum data. (If the machine sharks, I'll lower the limits...)
- test the peeler with its increased limits (in the published firmware the peeler runs much slower then X and hence limits the dragging)
- running the peeler as axis in parallel with the dragging motion might not be configured yet
- I only used and tested the published version on a 36VA. IIRC there are some modifications needed to operate the second peeler as axis as well.

Jan

PS: I'll try to get the review and some basic tests done in the next days.

vespaman commented:
Hi Jan,
>On the receive side (serial -> USB) the situation is much more complicated as for good throughput data shall be buffered and for low latency a end-of-message indication is needed. FTDI and IMHO the USB-ACM specification use timeouts and trigger character for that. An additional but related problem is given by the host-oriented nature of the USB: the host controller polls the bridge for new data and the polling rate can not be faster then 1ms. IIRC the generic Windows drivers allows to configure this interval and defaults to 10ms. I have not yet studied the datasheets of the bridges you mentioned but I would recommend one that supports the trigger-character and use a rather large timeout.

Well (IMHO), there's really no need to make it so complicated, the XR series does deal with these issues very well, and also a "real UART" would (not that it is very common to use real UARTs these days, though).

> Concerning the machine.xml: there are a few optimization, that are not yet in the published version:
> 
>     * test the peeler with its increased limits (in the published firmware the peeler runs much slower then X and hence limits the dragging)

Yes, this is also in my config, and I have been running with much higher speed on the peelers the last couple of months.

>     * I only used and tested the published version on a 36VA. IIRC there are some modifications needed to operate the second peeler as axis as well.

Yes, the main thing is to add the extra axis it to the config file, and in the machine.xml change the feed actuators into Profile actuators, and in the Profile actuators add the Dragpin actuator, this way they both can use the same drag pin properly.

vespaman commented:
Hi Dominic,
> yes, I would like a machine.xml which has config for the right hand feeders, and also for the drag pin management differences of your firmware. If you have yours that would be most helpful.

I have now checked in my machine.xml (and fixed the issues you found). Use this as a reference, but also understand that you cannot use it as is (because of cameras, calibrations and various setup differences).
One thing I should highlight already is that I have changed my nozzles into free turning (full articulation), so you will have will have to change those settings back on a stock machine.

janm012012 commented:
> Hi Jan,
> 
> > On the receive side (serial -> USB) the situation is much more complicated as for good throughput data shall be buffered and for low latency a end-of-message indication is needed. FTDI and IMHO the USB-ACM specification use timeouts and trigger character for that. An additional but related problem is given by the host-oriented nature of the USB: the host controller polls the bridge for new data and the polling rate can not be faster then 1ms. IIRC the generic Windows drivers allows to configure this interval and defaults to 10ms. I have not yet studied the datasheets of the bridges you mentioned but I would recommend one that supports the trigger-character and use a rather large timeout.
> 
> Well (IMHO), there's really no need to make it so complicated, the XR series does deal with these issues very well, and also a "real UART" would (not that it is very common to use real UARTs these days, though).
> 
I thinks, that was the result of your previous discussions as well: to me the latency (and its jitter) between e.g. location confirmation and next move matters as it happens after every coordination (suppose location confirmation is enabled). I checked the datasheet of the XR21V1410: the latency is three characters (+ USB polling) and can be configured to 0, but I have not seen an option to trigger on a specific character. In the datasheet of the PL2303, which seems to be quite common today and is in my bridge delivered with the 36VA, the datasheet does not specify how receptions trigger USB activity. However, there was this configuration message in data stream I analysed, so both might actually support it because its part of the CDC they claim to follow.

Jan

vespaman commented:
> I thinks, that was the result of your previous discussions as well: to me the latency (and its jitter) between e.g. location confirmation and next move matters as it happens after every coordination (suppose location confirmation is enabled). I checked the datasheet of the XR21V1410: the latency is three characters (+ USB polling) and can be configured to 0, but I have not seen an option to trigger on a specific character. In the datasheet of the PL2303, which seems to be quite common today and is in my bridge delivered with the 36VA, the datasheet does not specify how receptions trigger USB activity. However, there was this configuration message in data stream I analysed, so both might actually support it because its part of the CDC they claim to follow.
> 
> Jan

Hi Jan, 
Yes, I tested the PL2303 also. Infact I had two versions of it. It is a budget device, which is not too bad. But there are different versions of the 2303 around, and there's also a lot of counterfeits. I got the feeling that it was more or less EOL. There are now many even lower cost devices available, so maybe this is why. I also tested Cypress parts, Si parts, FTDI of course and so on. I kind of doubt this control character is available in any of the chips I tested, but may rather done in CDC driver in this case. Of the two devices I tested, only one did actually work well at 921kbps, the other looked like sh*t, and I ended throwing it away. Maybe it was a counterfeit one, or maybe a different version of the chip. 

But apart from this, if you take a few steps back; if you where to add this "end of message" character, and somehow add support for this in the PC end, the only thing you would achieve over "just using" one of the Exar devices "as is", is 2 character times. This is 176us at 115200. If you then compare this with the length of the actual message, the 176us is not really significant, especially if you also say windows has 1ms polling time. The confirmation message, is that not 20-25 bytes by itself?
Maybe it would be better to use a user defined baud rate for you - even a slightly higher baud rate will give you much better performance. The RS232 level chip on the chmt are only spec'd to 115k, but I'm sure you could go a little bit higher. Maybe even up to 140kbit? You would have to test, of course. 

But this is just my two cents. :-) 
Regards, 
 Micael

hydra commented:
> > yes, I would like a machine.xml which has config for the right hand feeders, and also for the drag pin management differences of your firmware. If you have yours that would be most helpful.
> 
> I have now checked in my machine.xml (and fixed the issues you found). Use this as a reference, but also understand that you cannot use it as is (because of cameras, calibrations and various setup differences). One thing I should highlight already is that I have changed my nozzles into free turning (full articulation), so you will have will have to change those settings back on a stock machine.

@vespaman Thanks for making it available, I'll use it as a reference as-needed.  I also just tried out the ASW and made a PR to this repo, see: https://github.com/janm012012/Smoothieware-CHMT/pull/2

janm012012 commented:
> I also just tried out the ASW and made a PR to this repo, see: #2

Many thanks, @hydra! That makes it way easier to merge this in.

If you'd like to continue, I'd suggest to break out fix to the serial transmitter code ;-) IIRC it fixes a concurrent problem between reception and transmission which is worked around by forcing "Confirmation Flow Control".

Jan

hydra commented:
> If you'd like to continue, I'd suggest to break out fix to the serial transmitter code ;-) IIRC it fixes a concurrent problem between reception and transmission which is worked around by forcing "Confirmation Flow Control".

Currently I'm only working on things that will make my machine into a usable machine again as it's just a giant paperweight right now.  Since there's a workaround the issue you mention is a nice-to-have and not a priority for me.  Perhaps @vespaman can make a more isolated PR if he's got more time?

janm012012 commented:
Hi @vespaman!

Would you please tell me that changes  you did to enhance the vacuum sensing? Unfortunately I did not found out by myself: you mention "vacuum" in three commits but only in commit 755b743 TemperaturControl.cpp is changes, which IMHO is where the vacuum sensor readings are coming from. However, the change is only minor (replaces "this" with "sensor") and I don't see how that effects the readout speed.

Jan

vespaman commented:
Hi @janm012012,
Yes, you found it! That is the  tiny little line, that makes the huge difference! :-) It will, IIRC change from 50ms to instant. Especially important if one is using vacuum for ending the pick dwell time. There will be some limitation sampling at stock speed, but not much, and way better than upstream Smoothieware.

  - Micael

janm012012 commented:
Hi @vespaman!

Tiny in deed! Many thanks!

I've now separated all features (vacuum sensing, hardware flow control and dma) and merged. Will need to make some more tests before I can release the result.

I found your hint, that on the 36VA mainboard the spare RS232 driver can be used to generate RTS. That sounds cool as it could be enabled without actual problems because if the wire is not present, the computer would just ignore the signal. However, it would result in two different hardware version which ideally could be selected using a configuration option. Do you think we should/can change the rts_cts_handshake property to reflect both? What would be most useful for new users?

Jan

vespaman commented:
Hi @janm012012,
Nice to hear! 
I think that might be possible to select different config, but I 'll have to have a look into it - it was some time since I did that code... The main issue is, that there are so many ways to solve things, and having both 48/36 did not help. But maybe, now, the dust has settled, and it can be simplified. It might be useful for me to revisit it, since I am considering changing RTS/CTS pins myself, since I possibly need the current pins for CAN bus.

janm012012 commented:
Hi @vespaman!

I've tested the merge today and even after fixing some rts == NC issues the code does not work on my CHM-T36VA mainboard with default serial interface. The controller runs again, but it does not respond at all. I already tried enabling the other serials in kernel.cpp with no success. Do you have any idea what's missing? Is the serial on the 36 using a different usart requiring a new/different DMA setup? If you'd like to have a look, I pushed my current version - including a big remark, that the code currently is dysfunctional.

Jan

vespaman commented:
Hi @janm012012
Can you be more specific, what exactly does not work? Does it receive, or is it tx that does not work? I have limited time to look into this right now, comparing the files gave a lot of formatting changes, so it is not so fun to follow.

janm012012 commented:
Hi  @vespaman!

I'm sorry, I felt I should adjust the formatting slightly to the surrounding code.

Actually I don't know what is not working. I complied the code with hardware handshake switch off and the controller did not start. After some code study, I found locations where a defined and configured RTS pin is silently assumes. After fixing that, the code works again. Next I tried detect firmware from within OpenPnP without success. I then studied the code again and found, that there are more serial consoles initialized in the original version. I enabled the again and still no response. Unfortunately I did not tested if rx or tx or both are not working.

First of, it would be great if you could verify that the code works for you. IIRC you'll have to set the baudrate back to 4mbit and enable hardware handshake. Then I'd expect success. If not, there is a merging issue.

IIRC the board exposes two usarts, one as RS232 and the other as RS422, which requires two serial consoles to support them both at once. Which one do you use?  If you use the RS232 one, its the same as I do, then there is a merging issue. If you're using the RS422 one, then we'll have to to verify the setup for the second usart.

What ever you can do, any help is welcome! I still don't have a debug setup for the controller. How did you set it up? (I've imported the project into Eclipse but did not linked/setup the build setup for debugging.)

Jan

vespaman commented:
I think the biggest no-no in git is to reformat code. That is just a waste of everyone's time. Now mine. 
But I have now gone through the merge, and it looks ok to my eye.

How did you test rx? You need to test with a command that is allowed even if the machine is halted (and, of course, a command that you can see works or not). You changed baud rate and hand shaking to disabled, and connected using some sort of serial terminal (also with handshaking disabled)?

janm012012 commented:
Hi @vespaman!

I reviewed, fixed and tested the code now: the 36VA variant uses UART1 while the 48VB uses UART2. Your modifications where all related to UART2. So I had to reenable UART1 and add dma support for it. Now it works as expected (without transmit buffer corruptions).

I've also documented the UART/serialConsole usage in kernel.cpp.

With a stock 48VB there is no RTS/CTS support on the RS422 link. Therefore the default has hardware flow control disabled and sticks with the 115.2mbaud. If hardware flow control is enabled, you get the setup you're using (UART1 with RTS and CTS). So if you at some point wont to go back to this branch, you'd have to enable hardware flow control and increase the baudrate. Everything else shall be the same then.

As diligent task one could join the dma stuff into a single handler functions taking the uart/obj as an argument. However, that's no so easy with ST's dma implementation as one would have to abstract all interrupt flags and register addresses as well. I think we shall keep it as is.

Many thanks for all your cool work!

Jan

janm012012 commented:
Hi @vespaman!

I've now modified rts_cts_handshake to an integer type and changed kernel.cpp to select between default (both UARTs without handshake, your and CHM-T36VA with RTS setup.

Your suggestion to use TX of the other UART as RTS on CHM-T36VA works. I've modified my machine accordingly.

I also have to revert my explanation for the required fix to get your code working for me: while reworking the rts_cts_handshake and isolating the serial actually used, I had to notice, that we are both using the same UART. So my mod is for the other UART that is likely used on stock 48VB mainboards via RS422 and hence untested yet. Maybe I can swap serial connectors on my mainboard to test it...

Jan

vespaman commented:
Hi @janm012012 

Good to hear!  I can promise you a much more stable implementation, and much less interrupt hits within the smoothie code.

 - Micael

janm012012 commented:
Hi @vespaman!

I'm trying to debug the reason why the controller works on UART1 but not on UART2 and found the POST_TX_WAIT define. Would you please tell me, why you decided to wait after TX rather then letting the DMA send the data and wait before the next TX? Is the calling code reusing the buffer while the DMA is still active?

Jan

vespaman commented:
Yes, I used to run without, this was the design I had chosen, but soon realized that if tx_after_ok, the caller will eventually reuse the same buffer (in stdlib or something like that), so I had to revert. And since I run with such a high baud rate, I never got the urge to fix it. This might be good to fix for you, since you are running at 115k2.

The simplest way is to deal with this in GcodeDispatch.cpp, where it is put in, probably by creating a temp buffer, that will be tested to be put into the serial queue on a callback. But this has to be thoroughly tested, since we need to make sure that the tx_after_ok is appended relatively quick and certainly before any other gcode is dispatched.

A even simpler way, not so nice though, would be to check the dma status when GcodeDispatch arrives with a tx_after_ok, and then and there wait for DMA to be finished before putting the new string into play. Probably good enough, but inconsistent, which I personally don't like.

janm012012 commented:
Hi @vespaman!

Many thanks! I've just added double-buffering using a private dma transmission buffer. That allows to safely send data in the back.

Jan

vespaman commented:
Hi Jan
That's another way of doing it. But now you get the memcpy penalty on all tx instead. This I would avoid personally (tx_after_ok is really not used often in comparison).

janm012012 commented:
Agreed! However, while studding the code I had the impression, that there is no way to control what memory is used. Initially I thought, I could provide two buffers there the stream is generated that ends up in serial_send_string(). But it seems, its routed via a 64 char buffer in StreamOutput.cpp located on the stack, which is replaced by a dynamic one if to short. This dynamic buffer is deleted immediately after the puts() it is hand over to. This looks to me, that puts() is designed to be blocking, which would require either your wait-while-dma-is-active or a copy as I'm now doing. I like my approach, because it allows to continue safely while the dma is active.
If I'm missing anything of if you see a way to safely get rid of the memcopy(), I'd be happy to read.

Jan

PS: Many thanks for all your help!

vespaman commented:
Yes, I don't think you can do it any other way, if you want to do it after it has been put into the stream. Maybe a DMA copy could be used, but I'm not sure about that. And it may not be more efficient anyway.

I would probably do it in GCodeDispatch, like I wrote. Then it is easy to just double buffer when needed, which is very seldom. Maybe even make the secondary buffer public, so the 'end of DMA' interrupt can pick it up and reload a new DMA tx as soon as the first DMA buffer has been emptied.
I know I also thought about changing the tx_after_ok concept altogether, but maybe it was not easy, or something else, that I have now forgotten. In fact, I think that using temp sensors for the vacuum is not very good, best would be to collect both vacuum  values and send back as one proper vacuum result, instead of the temp stuff, which complicates stuff. 
But unfortunately, also the ASW now uses tx_after_ok, so maybe it is not easy to get rid of it.. :-)

janm012012 commented:
Well. I've used memory-to-memory dma in other projects and they are difficult to setup and only useful if a lot of memory needs to be copied. I would also say, they are only useful if the CPU can do other work in parallel, but that requires even more setup work.

There is on think I might add to improve the situation: if the dma is running, pause it and try to append the new data. Don't know if that would work... Could be even extended into some kind of ringbuffer... Do you have any experience in that? And - main question - do you think it would be worth?

Jan

vespaman commented:
The 'problem' with tx_after_ok, is that it is a) not very common that it happens in chmt, and b) it is not much characters appended with tx_after_ok. So this is why a) I don't like the copy everytime solution much and b) setting up DMA for transferring the appendix is also not very efficient. 
Your thoughts on stopping the DMA - where would you do this, in what context? You can't do it in serial_send_string - it is too late. And stopping DMA from GCodeDispatcher will probably not be very clean from a coding perspective.

I *think* that the most efficient way is still to buffer the tx_after_ok data when it is around, and save pointer to that data that the dma interrupt can pick up (and clear the tx_after_ok).  <- Probably simple-ish

Or:- to remove tx_after_ok concept. (This is probably what I would do, and I think it is rather simple, at least with the vacuum stuff. Maybe the ASW reporting is trickier (don't remember off memory).  <- Hardest

Or still, wait for dma to finish if it is still ongoing when arriving to GCodeDispatcher with tx_after_ok. <- Simplest

Sorry if I repeat myself. :-) 
If you are going to mess with the setup DMA, be careful, and make sure you have some kind of test setup where you can push the system, it can be quite easy to miss (as you are probably aware of, if you have been working with DMA before), since it happens behind everyone's back, and it is hard to know at a given point in time where data is, and it is easy to loose a byte here or there.

Or:- you can leave it as you have done now, with copying everything into double buffer. In a way, with lower Baud rate, the copying is not as harmful, it is more harmful with higher Baud rates. And I remember the wast amount of data that is pushed in the "3'rd order simulation" stuff, something that I personally probably will want to pursue again, if I ever get the time...  But I'll stick to my fork anyway, since I  have other things on the horizon. So maybe, if you like/can accept the copying, this is 'good enough'?

  /- Micael

janm012012 commented:
Hi @vespaman!

I rechecked, the code:

- txt_after_ok seems to be no real problem because it is combined with "ok" into a single line (GcodeDispatcher.cpp line 406) before being send. I did not found any occasions where txt_after_ok generates a second printf() other then in case of error (GcodeDisplatcher.cpp line 383/385 and 389) which I have never seen.
- if stream->printf() as used all over in GcodeDispatcher.cpp is routed via puts() in StreamOutput.cpp (line 23), the DMA must not keep the buffer. Therefor your implementation to wait for end of DMA before returning is safe. My modification to copy the data into a dedicated buffer that is safe to access while the DMA is running, is also safe.

I understand that you don't like the memcpy() and prefer to wait for end of transmission. @4mbit thats a fair compromise. I personally prefer no to block insight serial_send_string(), which is more costly @115k2.

After reviewing the code, I think there is still room for improvements if reception goes on eg. while the controller is delaying its response to a M400 or G4 P[] command. If that queued data is processed, it will likely generate a serial of fast calls to serial_send_string() which are then limited by the serial link speed (in both implementations). To overcome this, serial_send_string() would need some sort of queue or ring-buffer to take the data without blocking. With the memcpy() code in place, one could disable a running DMA, append data to the buffer (suppose space permits) and reenable it. The usart would still continue transmitting providing an uninterrupted stream of data.

Em I missing anything?

Jan

vespaman commented:
Hi Jan,

Now I'm confused..

>txt_after_ok seems to be no real problem

In what sense? The problem I assumed we where talking about, is that this is the hole case for any sort of management of the buffers is because of tx_after_ok. 
If if was not for tx_after_ok, there would be absolutely no need to stop or do memcpy.. Do you agree with this?


 -  Micael

janm012012 commented:
Hi Micael!

Today - after my code study - I learned, that while processing a command, text can be added, to be send after the ok/error response. I assume, that is what you where referring to by "txt_after_ok".

IMHO there is a performance penalty, if serial_send_string() waits until all data has been send before returning. However, depending on the location of the input parameters, serial_send_string() must not read it after returning because its content my change. IMHO both of our implementations are safe with that respect.

For me, always waiting for transmission to finish, is wasting MCU power, especially for me running @115.2kbit. Therefor I prefer the copy and wait on start approach.

You then said, that there would be an issue with txt_after_ok and I assumed, that you have identified situations, where the ok response and the text to be send afterwords are send using two consecutive calls to serial_send_string(). If that's the case, waiting after the ok or before txt_after_ok makes no difference. However, according to my code studies, that seems not to happen.

And now we're here:

> The 'problem' with tx_after_ok, is that it is a) not very common that it happens in chmt, and b) it is not much characters appended with tx_after_ok.

This sounds like you expect, that ok and txt_after_ok are send separately.

However, as I don't know who is calling serial_send_string() exactly and having found this puts() in SerialStream.cpp. I think serial_send_string() has to be designed to not read the input after returning. I'd love to be corrected here...

I'm just wondering if you have seen any other pattern in the call to serial_send_string() what might be worth optimizing for.

Jan

vespaman commented:
Hi Jan,

>Today - after my code study - I learned, that while processing a command, text can be added, to be send after the ok/error response. I assume, that is what you where referring to by "txt_after_ok".

I explicitly mean gcode->txt_after_ok as is used at in GCodeDispatcher.
 
                        if(!gcode->txt_after_ok.empty()) {
                            new_message.stream->printf("ok %s\r\n", gcode->txt_after_ok.c_str());
                            gcode->txt_after_ok.clear();

This is the entire reason for the complication. Nothing else.

>IMHO there is a performance penalty, if serial_send_string() waits

Of course. But this is *only* because of tx_after_ok.

>This sounds like you expect, that ok and txt_after_ok are send separately.

They are. txt_after_ok is sent on another "round" in gcode_dispatcher. That is the whole purpose of tx_after_ok - to append data. This is happening on vacuum and also nowadays on ASW errors.


The stream->printf() is the starting point for the string being sent. 

>I think serial_send_string() has to be designed to not read the input after returning. I'd love to be corrected here...

The thing that needs to avoided, is to call stream->printf() before the DMA has done its peace.  This is *only* happening in the code above i.e.

                        if(!gcode->txt_after_ok.empty()) {
                            new_message.stream->printf("ok %s\r\n", gcode->txt_after_ok.c_str());

hence, my three suggestions again;
A check/wait can be done here; (this might look like a speed penalty but in reality it is really not)

                        if(!gcode->txt_after_ok.empty() ) {
                            while( dma_still_in_progress() ) ;
                            new_message.stream->printf("ok %s\r\n", gcode->txt_after_ok.c_str());


or like this 

                        if(!gcode->txt_after_ok.empty() ) {
                            snprintf( temp_buffer, ("ok %s\r\n", gcode->txt_after_ok.c_str());
                        .. and let serial_send_string() check if temp_buffer != null, in that case, start a new DMA tx with tempbuffer.

or simply change the vacuum code, and the ASW code, so they do not need txt_after_ok anymore. This should be simple with vacuum, not sure about ASW, it may need some alternative.


 - Micael

vespaman commented:
I should probably also be clear in that this test in GCodeDispatcher is done on two places, which you have probably seen;

if (gcode->is_error) {

and

}else if(!sent_ok) {

The second one is where we arrive with vacuum value for second nozzle.

  Micael

vespaman commented:
.. and OK, now I see that I have written tx_after_ok at some places in our communication. Guess my brain wasn't properly connected. So tx_after_ok == txt_after_ok.

Sorry about that

vespaman commented:
After thinking more about this, I was thinking about your double buffering, where do you do this, and how? 
Also, I specifically don't remember that the memory is freed after SerialConsole::puts(), are you sure about this? I have a vague memory of the buffer that is called by StreamOutput::printf() is not the same. But I may be wrong, many code lines has passed since I was fully on top of this. :)

janm012012 commented:
Hi Micael!

Double buffering for me is sending data in the background while processing more commands in the main loop keeping the com link busy and limiting the processing speed by com speed only. This requires serial_send_string() to not block at the end.

After a good night of sleep I found an other suspect: excessive use of printf(): IIRC each call to prinf() copies its input into a new buffer resisting in RAM. This buffer is in StreamOutput::printf() either "char b[64]", which resists on the stack or - if the 64 chars are not enough - "buffer = new char[size];" which resists on the heap and is deleted immediately after puts() returns. Both buffers are invalid when StreamOutput::printf() returns. Hence puts() either has to wait before returning or has to copy the data.

Looking into the code that calls StreamOutput::printf() I think it is primarily "new_message.stream->printf("ok\n");" in GCodeDisplatcher.cpp line 413. Here the input to printf() is constant and printf could be replaced by puts(). puts() is then called with a string resisting in flash (assuming that the compiler pools strings in flash, which they usually do). Strings in flash are known to be constant over runtime and hence do not have to be copied. With that knowledge puts() could reference the input directly without copy to RAM (the DMA on the F4 can access flash like ram). With this change, serial_send_string() could be changed to block while the dma is active, reference the input directly if ins in flash or copy to ram and block for transmission to complete if ins in RAM.
To my understanding there shall be only data in RAM if txt_after_ok is used. Then printf() is used to combine "ok" with the requested text into a single line. This shall happen only for M115 (detect firmware), M114 (position report) and M119 (read drag pin state). M15 is never used, M114 and M119 are always combined with coordination in OpenPnP so that there are no other commands the controller could process while sending the response. So there is no waist in processing power. This way we could combine your and my doubts into a single good solution.

What do you think?

Jan

vespaman commented:
Hi Jan,
>Double buffering for me is..

Yes, but where do you intend to do it? Or have you scrapped that idea? It might not be as easy as you think.
If you have scrapped it, then good. :-) 

Yes, I have also been thinking about the whole concept of the printfs, that are, like you say, just another copy in 90% of the cases. It could be handled better.  But I think you are mistaken regarding what is sent, and how. There's a lot of data sent with printfs, not just the line 413. Apart from vacuum and ASW, everything else is sent with "normal" printfs, all over the place. So I don't think that is a way forward.

However, this leads me to another solution, if you are sure the buffers in StreamOutput::printf() are what is delivered to serial_send_string(), then I think this might be good (enough);
Prerequisites:  Skip any double buffering you may have. Also remove the POST_TX_WAIT stuff.

Then, in StreamOutput::printf(), add two *static* buffers for ping-pong - pseudo code;

va_list args;

va_start(args, format);

b = NULL;

while(!b)

  if( !buffer0[0] )  {

    b = buffer0;

    size = vsnprintf(b, BUFFLEN, format, args) + 1;

  }

  else if (!buffer1[0])  {

    b = buffer1;

    size = vsnprintf(b, BUFFLEN, format, args) + 1;

  }

}

va_end(args);

puts(b);

return size -1;


Then, in the "end of DMA interrupt" clears the bufferX[0].

This way, we have double buffering without any extra cost, and also unlikely to be stopped regardless of txt_after_wait.

The only drawback, is that we add the static buffers, but I don't really see this as a problem. They probably don't have to be very big (64 each might be enough).

How does this sound to you?

How on earth does code formatting work in this editor?? 

 - Micael

janm012012 commented:
Hi Micael!

> Yes, but where do you intend to do it?

In my version I've implemented it: memcpy() + wait at entrance for end of running transfer.

> There's a lot of data sent with printfs, not just the line 413

Please enlight me where and which of them are of practical use for use. On my test run (for CommandStillstand) I have only seen M400, G1, M204, M119, M114 and the GCodes for drag pin and valve. Almost all of them generate "ok\n" as response and I assume via GCodeDispatcher.cpp line 403. This is the printf() I'd replace by a puts(). M114 and M119 generate more response. M114 is generated as txt_after_ok in Robot.cpp line 706. M119 is indeed a nightmare (again): it generates the result using multiple printfs(). (That reminds me, that I/we have to get rid of M119 to read the drag pin state or the drag pin interlock entirely... Btw: do you know of any other way to read a switch/input state via GCode?)

> if you are sure the buffers in StreamOutput::printf() are what is delivered to serial_send_string(),

Actually I don't know. To me it is easiest to use a debugger, break in serial_send_string() and read the call stack. I find reading and understanding call graphs generated by the compiler/linker especially from gcc much more difficult.

> There's a lot of data sent with printfs

Thats why I'll inspect every printf() and try to replace it with puts() in case its just a constant string. That will remove the copy done by printf().

> Then, in StreamOutput::printf(), add two _static_ buffers for ping-pong

The idea is good! (Keep in mind, that they need to be volatile for the compiler to always read the memory location and never optimize it away. Otherwise you might miss the signal from the IRQ.)

You could also extend you idea a little further and place every string on the heap. Then you could delete it once the transmission as finished. (I'm unsure if delete is safe to be called from interrupt context. However, it can be called after waiting for transfer to finish.) IIRC the dma retains the memory location.

Jan

vespaman commented:
Hi Jan,


OK, so you wait at entrance of serial_send_string() and once DMA is finished, you memcpy and continue as before? Yes that works.

So are you going to pinpoint only the gcodes that OpenPnP currenly uses? Is this really a meaningful way to solve the problem? If you do this, maybe you should also clean away all the dead code that is not used in current Smootihe, and do a Smoothie-lite just optimized for OpenPnP!

>Actually I don't know. To me it is easiest to use a debugger, break in serial_send_string() and read the call stack.

OK, you said earlier that it did. So I just wanted to make sure you had tested this. I don't want to do anything with my machine currently, since I am trying to build a stock. (I'm planning to unload all components and move the machine to another location, and at the same time I'll get rid of the noisy pump)

Heap mem can't be freed or allocated from interrupts. And it does not really add anything to do it on the heap anyway, since this function is called so often. But an improvement could be to let the end of DMA interrupt reload the other buffer into DMA, hence reducing possible waits further. But I don't really think it is needed with the current cpu utilization.

 - Micael

janm012012 commented:
Hi Micael!

You may have a look in my code. I just implemented to not copy data in flash anymore. This comes along side the printf() to puts() conversion where string are static.

In addition I added some compression to the GCode returned.

Jan
